File: src/cpp/ee/firebase_analytics/internal/FirebaseAnalyticsBridge.cpp
//
//  FirebaseAnalytics.cpp
//  ee_x
//
//  Created by Zinge on 5/10/17.
//
//

#include "ee/firebase_analytics/internal/FirebaseAnalyticsBridge.hpp"

#include <ee/core/ILogger.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/nlohmann/json.hpp>

namespace nlohmann {
template <typename... Args>
struct adl_serializer<std::variant<Args...>> {
    static void to_json(json& json, std::variant<Args...> const& value) {
        std::visit(
            [&json](auto&& v) { //
                json = std::forward<decltype(v)>(v);
            },
            value);
    }
};
} // namespace nlohmann

namespace ee {
namespace firebase {
namespace analytics {
namespace {
const std::string kPrefix = "FirebaseAnalyticsBridge";
const std::string kSetUserProperty = kPrefix + "SetUserProperty";
const std::string kTrackScreen = kPrefix + "TrackScreen";
const std::string kLogEvent = kPrefix + "LogEvent";
} // namespace

using Self = Bridge;

Self::Bridge(IMessageBridge& bridge, ILogger& logger,
             const Destroyer& destroyer)
    : bridge_(bridge)
    , logger_(logger)
    , destroyer_(destroyer) {}

Self::~Bridge() = default;

void Self::destroy() {
    destroyer_();
}

void Self::setUserProperty(const std::string& key, const std::string& value) {
    nlohmann::json request;
    request["key"] = key;
    request["value"] = value;
    bridge_.call(kSetUserProperty, request.dump());
}

void Self::trackScreen(const std::string& name) {
    bridge_.call(kTrackScreen, name);
}

void Self::logEvent(
    const std::string& name,
    const std::unordered_map<std::string,
                             std::variant<std::int64_t, double, std::string>>&
        parameters) {
    nlohmann::json request;
    request["name"] = name;
    request["parameters"] = parameters;
    bridge_.call(kLogEvent, request.dump());
}
} // namespace analytics
} // namespace firebase
} // namespace ee

File: src/android/firebase_analytics/src/main/java/com/ee/FirebaseAnalyticsBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import android.os.Bundle
import androidx.annotation.AnyThread
import com.ee.internal.deserialize
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.serialization.Contextual
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable

@InternalSerializationApi
class FirebaseAnalyticsBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val kTag = FirebaseAnalyticsBridge::class.java.name
        private const val kPrefix = "FirebaseAnalyticsBridge"
        private const val kSetUserProperty = "${kPrefix}SetUserProperty"
        private const val kTrackScreen = "${kPrefix}TrackScreen"
        private const val kLogEvent = "${kPrefix}LogEvent"
    }

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}

    override fun onDestroy() {
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(kSetUserProperty) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: String,
            )

            val request = deserialize<Request>(message)
            setUserProperty(request.key, request.value)
            ""
        }
        _bridge.registerHandler(kTrackScreen) { message ->
            trackScreen(message)
            ""
        }
        _bridge.registerHandler(kLogEvent) { message ->
            @Serializable
            class Request(
                val name: String,
                val parameters: Map<String, @Contextual FirebaseEventParameter>
            )

            val request = deserialize<Request>(message)
            logEvent(request.name, request.parameters)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kSetUserProperty)
        _bridge.deregisterHandler(kLogEvent)
        _bridge.deregisterHandler(kTrackScreen)
    }

    @AnyThread
    private fun setUserProperty(key: String, value: String) {
        Thread.runOnMainThread {
            Firebase.analytics.setUserProperty(key, value)
        }
    }

    @AnyThread
    private fun trackScreen(name: String) {
        Thread.runOnMainThread {
            Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
                param(FirebaseAnalytics.Param.SCREEN_NAME, name)
            }
        }
    }

    @AnyThread
    private fun logEvent(name: String, parameters: Map<String, FirebaseEventParameter>) {
        Thread.runOnMainThread {
            val bundle = Bundle()
            for ((key, value) in parameters) {
                if (value.type == FirebaseEventType.Long) {
                    bundle.putLong(key, value.longValue)
                }
                if (value.type == FirebaseEventType.Double) {
                    bundle.putDouble(key, value.doubleValue)
                }
                if (value.type == FirebaseEventType.String) {
                    bundle.putString(key, value.stringValue)
                }
            }
            Firebase.analytics.logEvent(name, bundle)
        }
    }
}